lib/services/publish-service.ts

```diff
@@ -0,0 +1,226 @@
+        return (() => {
+            var projectDir = this.$project.getProjectDir().wait();
+            var publishFilePath = path.join(projectDir, PublishService.JSON_PUBLISH_FILE_NAME);
+            if (!this.$fs.exists(publishFilePath).wait()) {
+                this.allPublishConnections = [];
+                return;
+            }
+
+            try {
+                this.allPublishConnections = this.$fs.readJson(publishFilePath).wait();
+            } catch (err) {
+                this.$errors.failWithoutHelp("The publish file %s is corrupted." + os.EOL +
+                        "Consider restoring an earlier version from your source control or backup." + os.EOL +
+                        "Additional technical information: %s", publishFilePath, err.toString());
+            }
+        }).future<void>()();
+    }
+
+    private getFtpConnectionData(idOrUrl: string, username: string, password: string): IFuture<Server.FtpConnectionData> {
+        return (() => {
+            var ftpConnectionData: Server.FtpConnectionData = {
+                RemoteUrl: idOrUrl,
+                ShouldPurge: options.force,
+                Username: username,
+                Password: password
+            };
+
+            if (!ftpConnectionData.Username) {
+                ftpConnectionData.Username = this.$prompter.getString("Username:").wait();
+            }
+
+            if (!ftpConnectionData.Password) {
+                ftpConnectionData.Password = this.$prompter.getPassword("Password:", {allowEmpty: true}).wait();
+            }
+
+            var id = parseInt(idOrUrl) - 1;
+
+            // the check whether it contains a dot is performed
+            // in case the user tries to publish to an IP address
+            if (!isNaN(id) && idOrUrl.indexOf('.') < 0) {
+                this.validateId(id);
+
+                ftpConnectionData.RemoteUrl = this.ftpPublishConnections[id].publishUrl;
+            }
+
+            return ftpConnectionData;
+        }).future<Server.FtpConnectionData>()();
+    }
+
+    private publishToRemote(ftpConnectionData: Server.FtpConnectionData): IFuture<void> {
+        return (() => {
+            var projectName = this.$project.projectData.ProjectName;
+            var projectDir = this.$project.getProjectDir().wait();
+            this.$projectFilesManager.excludeFile(projectDir, PublishService.JSON_PUBLISH_FILE_NAME);
+            this.$project.importProject().wait();
+            this.$logger.printInfoMessageOnSameLine("Publishing");
+
+            try {
+                this.$progressIndicator.showProgressIndicator(this.$server.publish.publishToFtp(projectName, projectName, ftpConnectionData), 6000).wait();
+            } catch (err) {
+                this.$logger.out();
+                if (err.message === PublishService.ERROR_MESSAGE_WHEN_PROJECT_EXISTS_ON_REMOTE) {
+                    this.$errors.failWithoutHelp("This project already exists on the remote server. " + os.EOL + 
```

Comment:
`this.$logger.out()` is needed because regardless of what error is thrown we don't want to print the message on the same line